kernel/src/hil/adc.rs

```diff
@@ -1,38 +1,80 @@
+/// Interfaces for analog to digital converter peripherals.
 use returncode::ReturnCode;
 
-/// Trait for handling callbacks from ADC module.
-pub trait Client {
-    /// Called when a sample is ready.
-    fn sample_done(&self, sample: u16);
-}
 
-/// Simple interface for reading a single ADC sample on any channel.
-pub trait AdcSingle {
+// *** Interfaces for low-speed, single-sample ADCs ***
+
+/// Simple interface for reading an ADC sample on any channel.
+pub trait Adc {
+    /// The chip-dependent type of an ADC channel.
+    type Channel;
+
     /// Initialize must be called before taking a sample.
-    /// Returns true on success.
     fn initialize(&self) -> ReturnCode;
 
     /// Request a single ADC sample on a particular channel.
-    /// Returns true on success.
-    fn sample(&self, channel: u8) -> ReturnCode;
-    fn cancel_sample(&self) -> ReturnCode;
+    /// Used for individual samples that have no timing requirements.
+    fn sample(&self, channel: &Self::Channel) -> ReturnCode;
+
+    /// Request repeated ADC samples on a particular channel.
+    /// Callbacks will occur at the given frequency with low jitter and can be set to any frequency
+    /// supported by the chip implementation. However callbacks may be limited based on how quickly
+    /// the system can service individual samples, leading to missed samples at high frequencies.
+    fn sample_continuous(&self, channel: &Self::Channel, frequency: u32) -> ReturnCode;
+
+    /// Stop a sampling operation.
+    /// Can be used to stop simple or high-speed samples. No further callback will occur in simple
+    /// mode. The `sampling_complete` callback will occur in high-speed mode.
+    fn stop_sampling(&self) -> ReturnCode;
```

Comment:
In the current design in this PR, `stop_sampling always succeeds unless the ADC is not currently running. The current semantics of `stop_sampling` are that the ADC will be ready to use again by the time it returns. If the ADC is operating in high-speed mode, a `sampling_complete` callback will occur before the `stop_sampling` command returns in order to return buffers.

Once `stop_sampling` is called, no additional `sample_ready` or `samples_ready` callbacks should be able to occur. It's possible that the chips driver will receive another interrupt with data, but since `stop_sampling` cleans up state, that interrupt will be ignored and not cause a callback.

Does that sound like an acceptable interface to you? I'm in agreement that this could be documented more clearly, in the TRD when I update it would be a good place.